File: 0078-subsets/0078-subsets.py
class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        op = []
        self.solve(nums, 0, op, res)
        return res
    
    def solve(self, nums: List[int], start: int, op: List[int], res: List[List[int]]):
        if start == len(nums):
            res.append(op.copy())
            return
        
        self.solve(nums, start + 1, op, res)
        op.append(nums[start])
        self.solve(nums, start + 1, op, res)
        op.pop()
        

File: 0078-subsets/0078-subsets.cpp
class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> subsets;
        vector<int>subset;
        createSubsets(0, nums, subsets, subset);
        return subsets;
    }
    
    void createSubsets(int idx, vector<int> nums, vector<vector<int>>& subsets, vector<int>& subset) {
        if (idx>=nums.size()) {
            subsets.push_back(subset);
            return;
        }
        subset.push_back(nums[idx]);
        createSubsets(idx+1, nums, subsets, subset);
        subset.pop_back();
        createSubsets(idx+1, nums, subsets, subset);
    }
};